pallets/transaction-payment/src/lib.rs

```diff
@@ -258,6 +258,62 @@ where
     }
 }
 
+pub trait ChargeTxFee {
+    fn charge_fee(who: Signatory, len: u32, info: DispatchInfo) -> TransactionValidity;
+}
+
+impl<T: Trait> ChargeTxFee for Module<T> {
+    fn charge_fee(who: Signatory, len: u32, info: DispatchInfo) -> TransactionValidity {
+        let fee = if info.pays_fee {
+            let len = <BalanceOf<T>>::from(len);
+            let per_byte = T::TransactionByteFee::get();
+            let len_fee = per_byte.saturating_mul(len);
+
+            let weight_fee = {
+                // cap the weight to the maximum defined in runtime, otherwise it will be the `Bounded`
+                // maximum of its data type, which is not desired.
+                let capped_weight = info
+                    .weight
+                    .min(<T as frame_system::Trait>::MaximumBlockWeight::get());
+                T::WeightToFee::convert(capped_weight)
+            };
+
+            // the adjustable part of the fee
+            let adjustable_fee = len_fee.saturating_add(weight_fee);
+            let targeted_fee_adjustment = NextFeeMultiplier::get();
+            // adjusted_fee = adjustable_fee + (adjustable_fee * targeted_fee_adjustment)
+            let adjusted_fee =
+                targeted_fee_adjustment.saturated_multiply_accumulate(adjustable_fee);
+
+            let base_fee = T::TransactionBaseFee::get();
+            let final_fee = base_fee.saturating_add(adjusted_fee);
+
+            final_fee
+        } else {
+            Zero::zero()
+        };
+        let imbalance;
+        match who {
+            Signatory::Identity(did) => {
+                imbalance = T::Currency::withdraw_identity_balance(&did, fee)
```

Comment:
Will this be undone if the transaction fails later on?

Reply:
No, this must be called after we are sure that the transaction won't fail or when we want to charge the fee even if the transaction fails. The normal tx flow will be
- charge gas fee
- do checks to make sure tx is valid and allowed
- charge module fee using this function
- make state changes

Reply:
It would be great to somehow ensure that we adhere to this flow. At the moment there is only one place where `charge_fee` is used. As we add more, it's going to be less and less obvious to spot violations in this flow. Which is why we discussed something like `post_dispatch`.